vllm/model_executor/models/chatglm.py

```diff
@@ -0,0 +1,288 @@
+        self,
+        hidden_size: int,
+        intermediate_size: int,
+        hidden_act: str,
+    ):
+        super().__init__()
+        self.gate_up_proj = ColumnParallelLinear(hidden_size,
+                                                 2 * intermediate_size,
+                                                 bias=False,
+                                                 gather_output=False,
+                                                 perform_initialization=False)
+        self.down_proj = RowParallelLinear(intermediate_size,
+                                           hidden_size,
+                                           bias=False,
+                                           input_is_parallel=True,
+                                           perform_initialization=False)
+        if hidden_act != "silu":
+            raise ValueError(f"Unsupported activation: {hidden_act}. "
+                             "Only silu is supported for now.")
+        self.act_fn = SiluAndMul()
+
+    def forward(self, x):
+        gate_up, _ = self.gate_up_proj(x)
+        x = self.act_fn(gate_up)
+        x, _ = self.down_proj(x)
+        return x
+
+
+class ChatGLMAttention(nn.Module):
+    """Multi-headed attention from 'Attention Is All You Need' paper"""
+
+    def __init__(
+        self,
+        hidden_size: int,
+        num_heads: int,
+        kv_channels: int,
+        multi_query_attention: bool,
+        multi_query_group_num: int,
+    ):
+        super().__init__()
+        self.hidden_size = hidden_size
+        tp_size = get_tensor_model_parallel_world_size()
+        self.total_num_heads = num_heads
+        assert self.total_num_heads % tp_size == 0
+        self.num_heads = self.total_num_heads // tp_size
+
+        self.projection_size = kv_channels * num_heads
+        # Per attention head and per partition values.
+        self.hidden_size_per_attention_head = self.projection_size // num_heads
+        self.num_attention_heads_per_partition = num_heads
+
+        self.qkv_hidden_size = 3 * self.projection_size
+        if multi_query_attention:
+            self.num_multi_query_groups_per_partition = multi_query_group_num
+            self.qkv_hidden_size = (
+                    self.projection_size + 2 * self.hidden_size_per_attention_head * multi_query_group_num
+            )
+
+        self.head_dim = hidden_size // self.total_num_heads
+        self.scaling = self.head_dim**-0.5
+        # pylint: disable=invalid-name
+        self.qkv_proj = ColumnParallelLinear(
+            hidden_size,
+            self.qkv_hidden_size,
+            bias=False,
+            gather_output=False,
+            perform_initialization=False,
+        )
+        self.o_proj = RowParallelLinear(
+            self.projection_size,
+            hidden_size,
+            bias=False,
+            input_is_parallel=True,
+            perform_initialization=False,
+        )
+        self.attn = PagedAttentionWithRoPE(self.num_heads,
```

Comment:
Hi, @Jeffwan, chatglm uses position_encoding_2d in the attention_layer,I think hereis a little problem by just using `PagedAttentionWithRope`. https://huggingface.co/THUDM/chatglm-6b/blob/main/modeling_chatglm.py#L457

Reply:
@irasin Hi Sin, thanks for helping review. Do you like to co-author the PR? I can open the repo write permission to you.

Reply:
Hi, @Jeffwan I do not understand why there is no code about the special attention mask of ChatGLM. Could you please give us some explaination~